File: logcli/parser.py
"""JSON log parsing and normalization module."""

import json
import ipaddress
from datetime import datetime
from typing import Dict, Any, Optional, Tuple
from user_agents import parse as parse_user_agent

from .config import BOT_SIGNATURES


class LogParser:
    """Parses and normalizes JSON log entries."""
    
    def __init__(self):
        self.bot_signatures = {sig.lower() for sig in BOT_SIGNATURES}
    
    def parse_log_line(self, line: str) -> Optional[Dict[str, Any]]:
        """Parse a single JSON log line and normalize fields."""
        try:
            raw_log = json.loads(line.strip())
            return self.normalize_log(raw_log)
        except (json.JSONDecodeError, KeyError, ValueError) as e:
            # Skip invalid log lines
            return None
    
    def normalize_log(self, raw_log: Dict[str, Any]) -> Dict[str, Any]:
        """Normalize log entry fields for Nginx JSON format."""
        normalized = {}
        
        # Time parsing - Nginx uses 'time' field
        time_str = raw_log.get('time', '')
        normalized['timestamp'] = self._parse_timestamp(time_str)
        
        # IP address - Nginx uses 'remote_addr'
        ip_str = raw_log.get('remote_addr', '')
        normalized['ip'] = self._parse_ip(ip_str)
        
        # User agent parsing
        ua_str = raw_log.get('user_agent', '')
        normalized['user_agent'] = ua_str
        normalized['parsed_ua'] = self._parse_user_agent(ua_str)
        normalized['is_bot'] = self._is_bot(ua_str)
        
        # HTTP details - parse from 'request' field (e.g., "POST /graphql HTTP/1.1")
        request_str = raw_log.get('request', '')
        method, path = self._parse_request_string(request_str)
        normalized['method'] = method
        normalized['path'] = path
        
        normalized['status'] = int(raw_log.get('status', 0))
        normalized['referer'] = raw_log.get('referer', '')
        
        # Response details - Nginx specific fields
        normalized['response_time'] = float(raw_log.get('request_time', 0))
        normalized['bytes_sent'] = int(raw_log.get('body_bytes_sent', 0))
        
        # Geographic info - Nginx provides country
        normalized['country'] = raw_log.get('country', '')
        
        # Additional Nginx-specific fields
        normalized['host'] = raw_log.get('host', '')
        normalized['server_name'] = raw_log.get('server_name', '')
        normalized['handler'] = raw_log.get('handler', '')
        normalized['port'] = raw_log.get('port', '')
        normalized['ssl_protocol'] = raw_log.get('ssl_protocol', '')
        normalized['ssl_cipher'] = raw_log.get('ssl_cipher', '')
        normalized['remote_user'] = raw_log.get('remote_user', '')
        
        # Original raw log for reference
        normalized['raw'] = raw_log
        
        return normalized
    
    def _parse_request_string(self, request_str: str) -> Tuple[str, str]:
        """Parse HTTP request string like 'POST /graphql HTTP/1.1'."""
        if not request_str:
            return 'GET', '/'
            
        parts = request_str.split(' ')
        if len(parts) >= 2:
            method = parts[0]
            path = parts[1]
            return method, path
        
        return 'GET', '/'
    
    def _parse_timestamp(self, time_str: str) -> Optional[datetime]:
        """Parse timestamp from various formats."""
        if not time_str:
            return datetime.now()
            
        # Common timestamp formats
        formats = [
            '%Y-%m-%dT%H:%M:%S.%fZ',  # ISO format with microseconds
            '%Y-%m-%dT%H:%M:%SZ',     # ISO format
            '%Y-%m-%d %H:%M:%S',      # Standard format
            '%d/%b/%Y:%H:%M:%S %z',   # Apache/Nginx format
            '%Y-%m-%dT%H:%M:%S%z',    # ISO with timezone
        ]
        
        for fmt in formats:
            try:
                return datetime.strptime(time_str, fmt)
            except ValueError:
                continue
                
        # Try parsing as Unix timestamp
        try:
            return datetime.fromtimestamp(float(time_str))
        except (ValueError, TypeError):
            pass
            
        return datetime.now()
    
    def _parse_ip(self, ip_str: str) -> Optional[ipaddress.IPv4Address]:
        """Parse IP address."""
        if not ip_str or ip_str == '-':
            return None
            
        try:
            # Handle X-Forwarded-For format (take first IP)
            if ',' in ip_str:
                ip_str = ip_str.split(',')[0].strip()
            
            return ipaddress.ip_address(ip_str)
        except ValueError:
            return None
    
    def _parse_user_agent(self, ua_str: str) -> Dict[str, str]:
        """Parse user agent string."""
        if not ua_str or ua_str == '-':
            return {'browser': 'Unknown', 'os': 'Unknown', 'device': 'Unknown'}
        
        try:
            parsed = parse_user_agent(ua_str)
            return {
                'browser': f"{parsed.browser.family} {parsed.browser.version_string}",
                'os': f"{parsed.os.family} {parsed.os.version_string}",
                'device': parsed.device.family if parsed.device.family != 'Other' else 'Desktop'
            }
        except Exception:
            return {'browser': 'Unknown', 'os': 'Unknown', 'device': 'Unknown'}
    
    def _is_bot(self, ua_str: str) -> bool:
        """Detect if user agent is a bot."""
        if not ua_str:
            return False
            
        ua_lower = ua_str.lower()
        return any(bot_sig in ua_lower for bot_sig in self.bot_signatures)
    
    def get_status_category(self, status: int) -> str:
        """Categorize HTTP status code."""
        if 200 <= status < 300:
            return 'success'
        elif 300 <= status < 400:
            return 'redirect'
        elif 400 <= status < 500:
            return 'client_error'
        elif 500 <= status < 600:
            return 'server_error'
        else:
            return 'other'

File: internal/parser/parser.go
package parser

import (
	"encoding/json"
	"fmt"
	"net"
	"strings"
	"time"

	"github.com/hpowernl/hlogcli/internal/config"
	"github.com/hpowernl/hlogcli/pkg/models"
	"github.com/mssola/useragent"
)

// LogParser handles parsing and normalization of log entries
type LogParser struct {
	botSignatures map[string]bool
}

// NewLogParser creates a new log parser instance
func NewLogParser() *LogParser {
	return &LogParser{
		botSignatures: config.BotSignatures,
	}
}

// ParseLine parses a single JSON log line and returns a normalized log entry
func (p *LogParser) ParseLine(line string) (*models.LogEntry, error) {
	line = strings.TrimSpace(line)
	if line == "" {
		return nil, fmt.Errorf("empty line")
	}

	var raw map[string]interface{}
	if err := json.Unmarshal([]byte(line), &raw); err != nil {
		return nil, fmt.Errorf("failed to parse JSON: %w", err)
	}

	return p.NormalizeLog(raw), nil
}

// NormalizeLog normalizes a raw log entry to a standard format
func (p *LogParser) NormalizeLog(raw map[string]interface{}) *models.LogEntry {
	entry := &models.LogEntry{
		Raw: raw,
	}

	// Parse timestamp
	if timeStr, ok := raw["time"].(string); ok {
		entry.Timestamp = p.parseTimestamp(timeStr)
	} else {
		entry.Timestamp = time.Now()
	}

	// Parse IP address
	if ipStr, ok := raw["remote_addr"].(string); ok {
		entry.IP = p.parseIP(ipStr)
	}

	// Parse user agent
	if ua, ok := raw["user_agent"].(string); ok {
		entry.UserAgent = ua
		entry.ParsedUA = p.parseUserAgent(ua)
		entry.IsBot = p.isBot(ua)
	}

	// Parse HTTP request
	if requestStr, ok := raw["request"].(string); ok {
		method, path := p.parseRequestString(requestStr)
		entry.Method = method
		entry.Path = path
	}

	// Parse status code
	if status, ok := raw["status"].(string); ok {
		fmt.Sscanf(status, "%d", &entry.Status)
	} else if statusFloat, ok := raw["status"].(float64); ok {
		entry.Status = int(statusFloat)
	}

	// Parse referer
	if referer, ok := raw["referer"].(string); ok {
		entry.Referer = referer
	}

	// Parse response time
	if reqTime, ok := raw["request_time"].(string); ok {
		fmt.Sscanf(reqTime, "%f", &entry.ResponseTime)
	} else if reqTimeFloat, ok := raw["request_time"].(float64); ok {
		entry.ResponseTime = reqTimeFloat
	}

	// Parse bytes sent
	if bytes, ok := raw["body_bytes_sent"].(string); ok {
		fmt.Sscanf(bytes, "%d", &entry.BytesSent)
	} else if bytesFloat, ok := raw["body_bytes_sent"].(float64); ok {
		entry.BytesSent = int64(bytesFloat)
	}

	// Parse country
	if country, ok := raw["country"].(string); ok {
		entry.Country = country
	}

	// Parse additional Nginx fields
	if host, ok := raw["host"].(string); ok {
		entry.Host = host
	}
	if serverName, ok := raw["server_name"].(string); ok {
		entry.ServerName = serverName
	}
	if handler, ok := raw["handler"].(string); ok {
		entry.Handler = handler
	}
	if port, ok := raw["port"].(string); ok {
		entry.Port = port
	}
	if sslProto, ok := raw["ssl_protocol"].(string); ok {
		entry.SSLProtocol = sslProto
	}
	if sslCipher, ok := raw["ssl_cipher"].(string); ok {
		entry.SSLCipher = sslCipher
	}
	if remoteUser, ok := raw["remote_user"].(string); ok {
		entry.RemoteUser = remoteUser
	}

	return entry
}

// parseRequestString parses HTTP request string like "POST /graphql HTTP/1.1"
func (p *LogParser) parseRequestString(requestStr string) (method, path string) {
	if requestStr == "" {
		return "GET", "/"
	}

	parts := strings.Split(requestStr, " ")
	if len(parts) >= 2 {
		return parts[0], parts[1]
	}

	return "GET", "/"
}

// parseTimestamp parses timestamp from various formats
func (p *LogParser) parseTimestamp(timeStr string) time.Time {
	if timeStr == "" {
		return time.Now()
	}

	// Common timestamp formats
	formats := []string{
		time.RFC3339,                       // "2006-01-02T15:04:05Z07:00"
		"2006-01-02T15:04:05.999999Z",      // ISO with microseconds
		"2006-01-02T15:04:05Z",             // ISO format
		"2006-01-02 15:04:05",              // Standard format
		"02/Jan/2006:15:04:05 -0700",       // Apache/Nginx format
		"2006-01-02T15:04:05-07:00",        // ISO with timezone
		"2006-01-02T15:04:05.999999-07:00", // ISO with microseconds and timezone
	}

	for _, format := range formats {
		if t, err := time.Parse(format, timeStr); err == nil {
			return t
		}
	}

	// Try parsing as Unix timestamp
	var timestamp float64
	if _, err := fmt.Sscanf(timeStr, "%f", &timestamp); err == nil {
		return time.Unix(int64(timestamp), 0)
	}

	return time.Now()
}

// parseIP parses IP address from string
func (p *LogParser) parseIP(ipStr string) net.IP {
	if ipStr == "" || ipStr == "-" {
		return nil
	}

	// Handle X-Forwarded-For format (take first IP)
	if strings.Contains(ipStr, ",") {
		ipStr = strings.Split(ipStr, ",")[0]
		ipStr = strings.TrimSpace(ipStr)
	}

	ip := net.ParseIP(ipStr)
	return ip
}

// parseUserAgent parses user agent string
func (p *LogParser) parseUserAgent(uaStr string) *models.UserAgentInfo {
	if uaStr == "" || uaStr == "-" {
		return &models.UserAgentInfo{
			Browser: "Unknown",
			OS:      "Unknown",
			Device:  "Unknown",
		}
	}

	ua := useragent.New(uaStr)

	browser, version := ua.Browser()
	browserStr := browser
	if version != "" {
		browserStr = fmt.Sprintf("%s %s", browser, version)
	}

	osInfo := ua.OS()
	device := "Desktop"
	if ua.Mobile() {
		device = "Mobile"
	} else if ua.Bot() {
		device = "Bot"
	}

	return &models.UserAgentInfo{
		Browser: browserStr,
		OS:      osInfo,
		Device:  device,
	}
}

// isBot detects if user agent is a bot
func (p *LogParser) isBot(uaStr string) bool {
	if uaStr == "" {
		return false
	}

	uaLower := strings.ToLower(uaStr)
	for signature := range p.botSignatures {
		if strings.Contains(uaLower, signature) {
			return true
		}
	}

	return false
}

// GetStatusCategory categorizes HTTP status code
func (p *LogParser) GetStatusCategory(status int) string {
	switch {
	case status >= 200 && status < 300:
		return "success"
	case status >= 300 && status < 400:
		return "redirect"
	case status >= 400 && status < 500:
		return "client_error"
	case status >= 500 && status < 600:
		return "server_error"
	default:
		return "other"
	}
}

// IsErrorStatus checks if a status code is an error
func (p *LogParser) IsErrorStatus(status int) bool {
	return status >= 400
}

// IsSuccessStatus checks if a status code is successful
func (p *LogParser) IsSuccessStatus(status int) bool {
	return status >= 200 && status < 300
}
